tooling/moshkt/compiler/src/main/kotlin/io/sweers/moshkt/compiler/MoshKtProcessor.kt

```diff
@@ -313,16 +333,38 @@ private data class Adapter(
     val hasCompanionObject: Boolean,
     val visibility: Visibility) {
   fun generate(adapterName: String, fileSpecBuilder: FileSpec.Builder) {
+    val nameAllocator = NameAllocator()
+    fun String.allocate() = nameAllocator.newName(this)
+
     val originalTypeName = originalElement.asType().asTypeName() as ClassName
-    val moshiParam = ParameterSpec.builder("moshi", Moshi::class.asClassName()).build()
-    val moshiProperty = PropertySpec.builder("moshi", Moshi::class.asClassName(), PRIVATE)
-        .initializer("%N", moshiParam)
-        .build()
-    val reader = ParameterSpec.builder("reader", JsonReader::class.asClassName()).build()
-    val writer = ParameterSpec.builder("writer", JsonWriter::class.asClassName()).build()
-    val value = ParameterSpec.builder("value", originalTypeName.asNullable()).build()
+    val moshiName = "moshi".allocate()
+    val moshiParam = ParameterSpec.builder(moshiName, Moshi::class.asClassName()).build()
+    val reader = ParameterSpec.builder("reader".allocate(),
+        JsonReader::class.asClassName()).build()
+    val writer = ParameterSpec.builder("writer".allocate(),
+        JsonWriter::class.asClassName()).build()
+    val value = ParameterSpec.builder("value".allocate(),
+        originalTypeName.asNullable()).build()
     val jsonAdapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asClassName(),
         originalTypeName)
+
+    // Create fields
+    val adapterProperties = propertyList
+        .map { it.typeName }
+        .distinct()
+        .associate { typeName ->
+          val propertyName = "${typeName.simplifiedName().allocate()}_Adapter"
+          val adapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asTypeName(), typeName)
+          typeName to PropertySpec.builder(propertyName, adapterTypeName, PRIVATE)
+              .initializer("%N.adapter%L(%L)",
+                  moshiParam,
+                  if (typeName is ClassName) "" else CodeBlock.of("<%T>", typeName),
+                  typeName.makeType())
+              .build()
+        }
+
+    val allocatedNames = propertyList.associate { it to it.name.allocate() }
```

Comment:
This was a little confusing, and might be clearer to just create `PropertySpec`s here for the local properties. Then you'd have something like `adapterProperties` and `localProperties` which would make the intention clear, and you could use `%N` instead of `var ${allocatedNames[prop]}: %T` all over the place.

Reply:
can local variables be made with `PropertySpec`s? I haven't tried, but I assumed not 🤔